lib/src/take_until.dart

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+/// Emits values from the stream until [trigger] fires.
+///
+/// Completing [trigger] differs from canceling a subscription in that values
+/// which are emitted before the trigger, but have further asynchronous delays
+/// in transformations following the takeUtil, will still go through. Cancelling
+/// a subscription immediately stops values.
+StreamTransformer<T, T> takeUntil<T>(Future trigger) => new _TakeUntil(trigger);
+
+class _TakeUntil<T> implements StreamTransformer<T, T> {
+  final Future _trigger;
+
+  _TakeUntil(this._trigger);
+
+  @override
+  Stream<T> bind(Stream<T> values) {
+    var controller = values.isBroadcast
+        ? new StreamController<T>.broadcast(sync: true)
+        : new StreamController<T>(sync: true);
+
+    StreamSubscription subscription;
+    var isDone = false;
+    _trigger.then((_) {
+      if (isDone) return;
+      isDone = true;
+      subscription?.cancel();
+      controller.close();
+    });
+
+    controller.onListen = () {
+      if (subscription != null || isDone) return;
```

Comment:
I'm not sure it's worth checking and throwing - but you're right that it shouldn't happen. This turns out to be a redundant check so I removed it.